Add xnnpack global average pool op

axitkhurana opened the pull request:
Summary:

Adaptive average pool with output size (1, 1) is a global average pool
For mobile use xnnpack to speed up that path

Test Plan:

pytest test/test_xnnpack_integration.py::TestXNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #{issue number}

kimishpatel commented on aten/src/ATen/native/AdaptiveAveragePooling.cpp:
```diff
@@ -101,6 +102,10 @@ namespace {
     if (!input.is_quantized() && output_size[0] == 1 && output_size[1] == 1) {
       // in this case, adaptive pooling is just computing mean over hw
       // dimensions, which can be done more efficiently
+      #if defined(C10_MOBILE) && defined(USE_XNNPACK)
```
Do add comment as to why we are doing special path here for future context.

kimishpatel commented on aten/src/ATen/native/xnnpack/GlobalAveragePooling.cpp:
```diff
@@ -0,0 +1,75 @@
+#ifdef USE_XNNPACK
```
Can you name the file AveragePooling with GlobalAveragePool just a special case?

kimishpatel commented on aten/src/ATen/native/xnnpack/GlobalAveragePooling.cpp:
```diff
@@ -0,0 +1,75 @@
+    xnn_status_success == run_status,
+    "xnn_setup_global_average_pooling_nwc_f32 failed!");
+
+  return output;
```
Always return output in the same format as input like here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/xnnpack/Convolution.cpp#L380. Thats what we try to adhere to.

kimishpatel commented on aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <ATen/native/xnnpack/Convolution.h>
 #include <ATen/native/xnnpack/Linear.h>
 #include <ATen/native/xnnpack/OpContext.h>
+#include <ATen/native/xnnpack/Engine.h>
```
WHy this include?

kimishpatel commented on aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:
```diff
@@ -88,6 +90,7 @@ TORCH_LIBRARY_IMPL(prepacked, CPU, m) {
   m.impl(TORCH_SELECTIVE_NAME("prepacked::conv2d_transpose_clamp_prepack"), TORCH_FN(createConv2dTransposeClampPrePackOpContext));
   m.impl(TORCH_SELECTIVE_NAME("prepacked::conv2d_clamp_run"), TORCH_FN(internal::convolution2d::conv2d_clamp_run));
   m.impl(TORCH_SELECTIVE_NAME("prepacked::conv2d_transpose_clamp_run"), TORCH_FN(internal::convolution2d::conv2d_transpose_clamp_run));
+  m.impl(TORCH_SELECTIVE_NAME("prepacked::global_average_pool"), TORCH_FN(global_average_pool));
```
Same comment as previous diff. See if it is possible.

kimishpatel reviewed (changes requested):
Looks good. Minor changes.

Commit:
Add inplace hardswish xnnpack op (#56715)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/56715

Pull Request resolved: https://github.com/pytorch/pytorch/pull/55801

Refactor to add inplace version of xnnpack hardswish op

Test Plan: buck test //xplat/caffe2:pt_xnnpack_test

Differential Revision: D27712305

fbshipit-source-id: 69b53efedbd9991a5a119121a4292a1b19fbb19b

Commit:
Add xnnpack hardswish op (#56714)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/56714

Pull Request resolved: https://github.com/pytorch/pytorch/pull/55800

For mobile use xnnpack implementation of hardswish

Test Plan: buck test //xplat/caffe2:pt_xnnpack_test

Differential Revision: D27712306

fbshipit-source-id: 82fa5615f0437fd4f5898be7d96f7d8d74c00806

Commit:
Add xnnpack global average pool op (#55791)

Summary:
Adaptive average pool with output size (1, 1) is a global average pool
For mobile use xnnpack to speed up that path

Pull Request resolved: https://github.com/pytorch/pytorch/pull/55791

Test Plan:
buck test //xplat/caffe2:pt_xnnpack_test

pytest test/test_xnnpack_integration.py::TestXNNPACKOps
Fixes #{issue number}

Reviewed By: kimishpatel

Differential Revision: D27711082

Pulled By: axitkhurana

fbshipit-source-id: 594d54874db16c212e3406c9104949a8914cf015
